.github/workflows/demo-testing.yml

```diff
@@ -14,15 +14,23 @@ jobs:
       - name: build demo image
         run: podman build -f demo.Dockerfile -t localhost/testing:latest
 
+      - name: install yq for testing
+        run: go install github.com/mikefarah/yq/v4@latest
+
       - name: run demo image and ensure violations output unchanged
         run: |
           podman run -v $(pwd)/demo-output.yaml:/analyzer-lsp/output.yaml:Z localhost/testing:latest
-          git diff --exit-code demo-output.yaml
+          diff \
+            <(sort <(sed 's/incidents\.[0-9]\+/incidents\.x/g' <(yq -P 'sort_keys(..)' -o=props <(git show HEAD:demo-output.yaml)))) \
```

Comment:
what sorcery is this?

Reply:
Since the incidents in a violation can be returned in any order, `git diff` kept failing. i.e. `git diff` would compare

```yaml
sample-rule:
  incidents:
  - A
  - B
```
```yaml
sample-rule:
  incidents:
  - B
  - A
```

and fail This is because diff only compares the _textual differences_ between the files. It has no knowledge of the YAML structure. To fix this, I installed [`yq`](https://mikefarah.gitbook.io/yq) which allows us to examine the actual structure the input YAML. My whole goal was to still use `diff` to compare the two files, but make the position of each incident in the array irrelevant.

Moving from right to left, I first get the different versions of `demo-output.yaml`. I then pipe that into `yq`, with the [`-o=props` flag ](https://mikefarah.gitbook.io/yq/usage/properties).  `-P 'sort_keys(..)'` is actually unnecessary due to what I do later. If we piped the YAML files I showed into that, it would return:

```yaml
sample-rule.incidents.0 = A
sample-rule.incidents.1 = B
```
```yaml
sample-rule.incidents.0 = B
sample-rule.incidents.1 = A
```

I then pipe that output into `sed`, replacing every `incidents.<number>` with `incidents.x` to erase which index each element is in. Finally, I sort the result and `diff` the two streams. Both files would return

```yaml
sample-rule.incidents.x = A
sample-rule.incidents.x = B
```

Which is the same thing. No error, ta-da!

There might be a better way than cobbling together some shell commands for this. I wouldn't even be surprised if there was a way you could do all of this in `yq`! At least one of the advantages of having it installed is that we can investigate the actual structure of the YAML now in our tests.

Reply:
@JonahSussman lovely, I think this is a much better version of what we are doing in Kantra now, would love to copy it there

Reply:
I would consider it a great honor, copy away! I would remove the `-P 'sort_keys(..)'` part as I'm almost positive it's redundant.

Reply:
Can we sort these in the engine then?